File: java/composite.java
import java.util.ArrayList;
import java.util.List;

abstract class ElementoMenu {
    public abstract void mostrar(int nivel);
}

class Accion extends ElementoMenu {
    private String nombre;

    public Accion(String nombre) {
        this.nombre = nombre;
    }

    @Override
    public void mostrar(int nivel) {
        System.out.println("  ".repeat(nivel) + "Acción: " + nombre);
    }
}

class Menu extends ElementoMenu {
    private String nombre;
    private List<ElementoMenu> elementos = new ArrayList<>();

    public Menu(String nombre) {
        this.nombre = nombre;
    }

    public void agregar(ElementoMenu elemento) {
        elementos.add(elemento);
    }

    @Override
    public void mostrar(int nivel) {
        System.out.println("  ".repeat(nivel) + "Menú: " + nombre);
        for (ElementoMenu elem : elementos) {
            elem.mostrar(nivel + 1);
        }
    }
}

public class composite {
    public static void main(String[] args) {
        Menu menuPrincipal = new Menu("Archivo");

        Menu nuevoMenu = new Menu("Nuevo");
        nuevoMenu.agregar(new Accion("Documento"));
        nuevoMenu.agregar(new Accion("Carpeta"));

        menuPrincipal.agregar(nuevoMenu);
        menuPrincipal.agregar(new Accion("Abrir"));
        menuPrincipal.agregar(new Accion("Guardar"));

        Menu menuEdicion = new Menu("Edición");
        menuEdicion.agregar(new Accion("Cortar"));
        menuEdicion.agregar(new Accion("Pegar"));

        Menu root = new Menu("Aplicación");
        root.agregar(menuPrincipal);
        root.agregar(menuEdicion);

        System.out.println("Menú de la Aplicación:");
        root.mostrar(0);
    }
}

File: composite.py
from abc import ABC, abstractmethod

# Componente base
class ElementoMenu(ABC):
    @abstractmethod
    def mostrar(self, nivel=0):
        pass

# Hoja: Acción individual
class Accion(ElementoMenu):
    def __init__(self, nombre):
        self.nombre = nombre

    def mostrar(self, nivel=0):
        print("  " * nivel + f"Acción: {self.nombre}")

# Compuesto: Menú que contiene otros elementos
class Menu(ElementoMenu):
    def __init__(self, nombre):
        self.nombre = nombre
        self.elementos = []

    def agregar(self, elemento: ElementoMenu):
        self.elementos.append(elemento)

    def mostrar(self, nivel=0):
        print("  " * nivel + f"Menú: {self.nombre}")
        for elem in self.elementos:
            elem.mostrar(nivel + 1)


menu_principal = Menu("Archivo")

nuevo_menu = Menu("Nuevo")
nuevo_menu.agregar(Accion("Documento"))
nuevo_menu.agregar(Accion("Carpeta"))

menu_principal.agregar(nuevo_menu)
menu_principal.agregar(Accion("Abrir"))
menu_principal.agregar(Accion("Guardar"))

menu_edicion = Menu("Edición")
menu_edicion.agregar(Accion("Cortar"))
menu_edicion.agregar(Accion("Pegar"))

root = Menu("Aplicación")
root.agregar(menu_principal)
root.agregar(menu_edicion)

print("Menú de la Aplicación:")
root.mostrar()